### ISSUE TITLE
Regression tests sometimes die in test misc-2

### ISSUE THREAD

**Issue by kpoeck** [clasp-developers/clasp#1320]
**Describe the bug**
````
Passed MISC-1
../src/gctools/hardErrors.cc:46 Bad client pointer 0x7ff7b79e96e0
FAILED: test-boehmprecise 
boehmprecise/iclasp-boehmprecise --norc --non-interactive --feature ignore-extensions --load "sys:src;lisp;regression-tests;run-all.lisp"
ninja: build stopped: subcommand failed.
```` 
**Code at issue**

```lisp
(test-true misc-2
           (> (length (with-output-to-string (*standard-output*)
		        (room)))
              0))
```

**Other steps to reproduce**
````
ninja -C build test  
````
* Happened once to me on macos
* Has happened sometimes in the cI-runner on github
* comment on irc perhaps relevant
````
[18:41:30](https://irclog.tymoon.eu/libera/%23clasp?around=1653158490#1653158490)
drmeister
Ok - I found something.
[18:41:44](https://irclog.tymoon.eu/libera/%23clasp?around=1653158504#1653158504)
drmeister
ROOM is walking a BlockDynEnv_O
[18:42:23](https://irclog.tymoon.eu/libera/%23clasp?around=1653158543#1653158543)
Bike
ah yeah, those may retain references to deallocated stack objects
[18:42:31](https://irclog.tymoon.eu/libera/%23clasp?around=1653158551#1653158551)
Bike
i was a bit worried about how that could affect the GC
````
**Context**
````
git log ->
commit 561d0b4093b39e320000f6468820c3505c1a3c67 (HEAD -> main, upstream/main)
Author: Christian Schafmeister <meister@temple.edu>
Date:   Sat May 21 13:53:41 2022 -0400

    Improved ROOM function
    
    It uses its argument now and it
    writes everything to a std::ostringstream
    before writing it out to *standard-output*.
````

**Comment by drmeister** [comment]
I am seeing hangs on macOS when I run ROOM.
I added some debugging code and found one of Bike's dynamic environments on the heap that was pointing to a invalid object on the stack.

I added the ability to stack allocate cons cells and Bike is going to change the dynamic environments so that they don't have a parent slot that can point to invalid objects on the stack.  Instead he is going to create a stack allocated linked list of dynamic environments, most will be on the stack and some will be on the heap.

Once this is done we can re-evaluate this test and see if it passes.  I suspect that the problem was one of these heap allocated dynamic environments.
